### PR TITLE
Issue272 + Issue265 mpi and main in separate process

### LINKED ISSUE #265: Tests wrapping `main` to be run in a separate process
## Feature Request

**Describe the new feature:**
We have seen a pattern of making tests out of the `main` function of applications or of integration examples or tests.  Currently, wrapping main in a test and calling it more than once violates the fundamental assumptions of writing a `main` function - i.e., that it is called once and only once.  This can potentially lead to runtime nondeterminism (not seen when running the application outside of the test), or even undefined behavior.

Before talking about the proposal, let me give a brief description of how this has been accomplished in the past.  We would follow a pattern like the following

```c++
#define main main_impl
#include "main.cpp"
#undef main

#include <flit.h>

#include <string>

template <typename T>
class MyTest : public flit::TestBase<T> {
public:
    ...
protected:
    virtual flit::Variant run_impl(const std::vector<T> &ti) override {
        // redirect standard out to a char buffer
        int argc = ...; // some value
        char* argv[] = {...}; // some command-line arguments
        int retval = main_impl(argc, argv);
        // convert the standard output into a std::string
        std::string buffer = ...;
        return buffer;
    }
};
...
```

Above, the `main` function is renamed to `main_impl` by using a `#define` before the `#include`, followed by a `#undef`.  Using this approach, we can call the (now renamed) main function inside of the test.

**Suggested change:**
The proposal is to separate out the call to `main` in a separate process.  This would be done by recursively calling the same executable (i.e., `argv[0]`), but in a different mode that only runs the one test one time and outputs the test information to a file.  The parent process then reads the file to get the test results.  This approach allows us to create as clean of an environment as we can conceivably create for the user's `main` function, and it will only be called once within a process.  We can still do timing because we can spawn a child process for each time we want to execute the test.

But, we do not want to take this approach for all FLiT tests.  For unit tests, creating separate processes is too much unnecessary overhead.  To this end, we believe we could have another test base class that would spawn another process, something like

```c++
template <typename T>
class SeparateProcessTestBase : public TestBase<T> {...};
```

This class would have the same pure virtual functions but would reimplement `run` to either spawn a separate process or to call `run_impl` and output to a file.  Whether that is in a separate function or not remains as an implementation detail.  It would be great if we could reuse `flit::TestBase::run()` from `flit::SeparateProcessTestBase::run()`.

Below is a workout Ian and I did on the whiteboard to handle this problem and also the problem of MPI, specifically addressing how we can have the test call `MPI_Init()` and `MPI_Finalize()`.  The MPI aspect is a separate issue that needs to be implemented at the same time this one is done, since doing one without the other breaks functionality.

flit-mpi-whiteboard-plan

**Alternative approaches:**
A complete rearchitecting of the testing framework could be done instead.  We could have a test runner either in C++ or in Python.  This kind of approach would probably have ALL tests be done as separate processes and everything is managed instead by a test runner.  We would probably still want all of the tests compiled into a single executable just so that disk usage is not unnecessarily consumed (think about 8 tests with 250 compilations, each one being 5 MB, that would be 10 GB of space - not cheap).

### LINKED ISSUE #272: MPI to be run from child processes
## Feature Request

**Describe the new feature:**
MPI is currently handled in FLiT by having FLiT call `MPI_Init()` and `MPI_Finalize()`, then the user can just write MPI code in their test.  However, this is severely limited.  Most use cases we have seen of writing a test around MPI code has been to wrap the test around their `main()` function, in which we have to stub out `MPI_Init()` and `MPI_Finalize()` (as well as potentially some others).

Furthermore, this can potentially change functionality since the user may be calling `MPI_Init()` with some modified command-line arguments.  Also, our current approach does not jive with the feature request in issue #265.

The new change is to support MPI, but only in a way that it is done through child processes, similar to issue #265.  This is to ensure that their MPI code works as expected and there can be no blame associated to the FLiT framework.

**Suggested change:**

- Remove the MPI functionality currently (except for the compiling phase)
- Add a new function called `call_mpi_main()` (or something similar) where the user can call the given function as a child process wrapped around `mpirun`, with a user-specified mpirun string (such as `"srun -n 2"` or `"mpirun -n 2"`).

This has the following benefits:

- we will not need to stub out anything, so no need to use pMPI
- user code will be executed as expected
- calls such as `terminate()`, `abort()`, `exit()`, or even `MPI_abort()` are totally fine to do, since it happens in a child process
- each test can use a different run configuration
- we can safely run the MPI code more than once to get reliable timing

The following downsides to this approach:

- all MPI code must be in a main-like function (taking `argc` and `argv`) that has been registered with FLiT.
- backward-compatibility breaks
- more overhead (performance and developer time) if you are doing unit testing on your MPI functions since along with each test, you will need to write a main-like function
- need to output your results either to standard out, standard error, or a file
- cannot use the test class template parameter easily in the main-like function (I was able to do it, but it is definitely not easy)

**Alternative approaches:**

Currently, cannot think of any...

### PR DESCRIPTION
**Pull request by mikebentley15**
Fixes #265 and fixes #272

**Description:**
Allow calling a `main()`-like function from a FLiT test.  Also, be able to call a `main()`-like function that uses MPI from a FLiT test.

The `main()`-like functionality is the following:

- Done with two components:
    1. `REGISTER_FLIT_MAIN()` that takes the function pointer, and 
    2. `call_main()` that takes the function pointer and command-line arguments
- The `main()`-like function is called from a child process (that is the same executable, but with special flags)

The MPI functionality is the following:

- Remove all MPI functionality from before (except for the compiling with MPI)
- When FLiT tests are called, they are not under MPI
- Instead, the user calls `call_mpi_main()` from the test where they specify everything the `call_main()` function requires plus an `mpirun` command with flags.
- The new subprocess(es) is(are) now launched with the user's `mpirun` command and everything just works.
- The downside is that tests using MPI **must** use a `main()`-like function.  There is no other way.  I think this is a reasonable downside, and worth the exxtra stability our current approach affords
- Add utility functions in `src/subprocess.cpp`

**Documentation:**

- Remove documentation to removed MPI functionality
- Add many sections to "Writing Test Cases" page describing how to write a test that wraps around a `main()` function and how to write MPI tests with examples.

**Tests:**

- Update the MPI tests
- Test everything in `src/subprocess.cpp`
- Add unit tests for `call_main()`
- Add tests for the `TestFactory` class
- Add tests for the `registerTest()` function used by the `REGISTER_TEST()` macro

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -121,7 +121,9 @@ matrix:
 
 before_install: pip3 install --user toml $extra_pip
 
-script: make -j4 && make -j4 -C tests && make check
+script:
+  - make -j4 && make -j4 -C tests && make check
+  - readelf --debug-dump=info src/InfoStream.o | head
```
Comment (IanBriggs) [inline .travis.yml]: This looks to be left over from some debugging

Comment (mikebentley15) [reply .travis.yml]: You're absolutely right.  It was left over from trying to find out what version of Dwarf is being used on TravisCI.  Unfortunately, it was the same version that is used on my laptop as well, so I don't yet know why my laptop fails and TravisCI does not.

I'll remove it.  Thank you.

File: src/fsutil.cpp [hunk 2]
```diff
@@ -198,29 +200,59 @@ void chdir(const std::string &directory) {
...
+  free(fullpath);
+  return fullpath_string;
 }
 
-TempDir::TempDir() {
-  char fname_buf[L_tmpnam];
-  char *s = std::tmpnam(fname_buf);    // gives a warning, but I'm not worried
-  if (s != fname_buf) {
-    throw std::ios_base::failure("Could not find temporary directory name");
+std::string which(const std::string &command, const std::string &path) {
```
Comment (IanBriggs) [inline src/fsutil.cpp]: I feel like this function is inconsistent.

which("/not-a-command", <path>) will return "/not-a-command"
but
which("not-a-command", <path>) will throw a failure

Same goes for relative paths.

I can add the checks, I just want feedback on whether this make sense.

Comment (mikebentley15) [reply src/fsutil.cpp]: Sure, we could do
```c++
if (command[0] == '/') {
  auto status = file_status(command);
  if (status.is_reg) {
    return command;
  }
}
```
does that sound good?

Comment (mikebentley15) [reply src/fsutil.cpp]: The `file_status()` throws a `std::ios_base::failure` if there is no such file or directory, or the current user does not have permission to see the file or folder.

Comment (mikebentley15) [reply src/fsutil.cpp]: I can address this

Comment (mikebentley15) [reply src/fsutil.cpp]: Thanks for the find.  Yes, this should be fixed.  I also found that I have no unit tests for the `which()` function.  That's a problem, too.  I'll do that now.

Comment (mikebentley15) [reply src/fsutil.cpp]: Unit tests added and the `which()` function has been changed to act as expected according to the unit tests.  It is much more consistent now, albeit much longer :dizzy:

Comment (IanBriggs) [reply src/fsutil.cpp]: Awesome! Thank you.

**Pull request merged**